crates/cpuid-utils/src/lib.rs

```diff
@@ -23,55 +63,475 @@ pub mod bits;
+    pub fn len(&self) -> usize {
+        let mut len = 0;
+        for sl in self.0.values() {
+            match sl {
+                Subleaves::Absent(_) => len += 1,
+                Subleaves::Present(sl_map) => len += sl_map.len(),
+            }
+        }
+
+        len
+    }
+
+    /// Returns an iterator over the ([`CpuidIdent`], [`CpuidValues`]) pairs in
+    /// the map.
+    pub fn iter(&self) -> CpuidMapIterator {
+        CpuidMapIterator::new(self)
+    }
+}
+
+/// An iterator over a [`CpuidMap`]'s leaf/subleaf/value tuples.
+pub struct CpuidMapIterator<'a> {
+    leaf_iter: btree_map::Iter<'a, u32, Subleaves>,
+    subleaf_iter: Option<(u32, btree_map::Iter<'a, u32, CpuidValues>)>,
+}
+
+impl<'a> CpuidMapIterator<'a> {
+    fn new(map: &'a CpuidMap) -> Self {
+        Self { leaf_iter: map.0.iter(), subleaf_iter: None }
+    }
+}
+
+impl Iterator for CpuidMapIterator<'_> {
+    type Item = (CpuidIdent, CpuidValues);
+
+    fn next(&mut self) -> Option<Self::Item> {
+        // If a subleaf iteration is in progress, try to advance that iterator.
+        // If that produces another subleaf value, return it. Otherwise, clear
+        // the subleaf iterator and move to the next leaf.
+        if let Some((leaf, subleaf_iter)) = &mut self.subleaf_iter {
+            if let Some((subleaf, value)) = subleaf_iter.next() {
+                return Some((CpuidIdent::subleaf(*leaf, *subleaf), *value));
+            };
+
+            self.subleaf_iter = None;
+        }
+
+        // Advance the leaf iterator. If there are no more leaves to iterate,
+        // the entire iteration is over.
+        let (leaf, subleaves) = self.leaf_iter.next()?;
+
+        // Iteration has moved to a new leaf. Consider whether it has any
+        // subleaves. If it doesn't, simply return the leaf and value.
+        // Otherwise, start iterating over subleaves.
+        match subleaves {
+            Subleaves::Absent(val) => Some((CpuidIdent::leaf(*leaf), *val)),
+            Subleaves::Present(sl_map) => {
+                let mut subleaf_iter = sl_map.iter();
+
+                // This invariant is upheld by insert/remove.
+                let (subleaf, value) = subleaf_iter
+                    .next()
+                    .expect("subleaf maps always have at least one entry");
+
+                // Stash the iterator along with the leaf value it
+                // corresponds to so that future iterations can return
+                // the entire leaf/subleaf pair.
+                self.subleaf_iter = Some((*leaf, subleaf_iter));
+                Some((CpuidIdent::subleaf(*leaf, *subleaf), *value))
+            }
+        }
+    }
+}
+
+/// A map from CPUID leaves to CPUID values that includes a vendor ID. Callers
+/// can use the vendor ID to interpret the meanings of any extended leaves
+/// present in the map.
+#[derive(Clone, Debug)]
+pub struct CpuidSet {
+    map: CpuidMap,
+    pub vendor: CpuidVendor,
```

Comment:
this being public rather than having an accessor allows code in other crates to mutate the vendor after constructing a map, which...is probably fine but _feels weird_? i suppose none of this type's methods here actually interprets the CPUID values based on the vendor ID, so it's not like any invariant could be violated due to reaching in and changing it from AMD to Intel or whatever...

Reply:
It turns out this is only mutated in one place--the `cpuid::Specializer` has a `with_vendor` option that will override the vendor specified in the input `CpuidSet`, but as far as I can tell this option is never used. It may have been meant for a world where a caller reads the host vendor and then configures the Specializer to override it, but with this change we've drifted toward just having callers set up the CpuidSet in the way they want from the get-go. In any case I've removed the `pub` and added a getter for this.